torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -1083,6 +1020,53 @@ void TensorPipeAgent::markFutureWithError(
   }
 }
 
+std::vector<c10::DeviceIndex> TensorPipeAgent::getDevicesForTensors(
+    const std::string& remoteName,
+    const Message& message) const {
+  const auto& deviceMaps =
+      message.isRequest() ? opts_.deviceMaps : reverseDeviceMaps_;
+
+  const auto errStr = c10::str(
+      "TensorPipe RPC backend only supports CPU tensors by default, please "
+      "move your tensors to CPU before sending them over RPC, or call "
+      "`set_device_map` on `TensorPipeRpcBackendOptions` to explicitly "
+      "configure device mapping. ",
+      message.isRequest() ? "Request" : "Response",
+      " device mapping is not available for destination ",
+      remoteName);
+
+  const auto& iter = deviceMaps.find(remoteName);
+  if (iter == deviceMaps.end()) {
+    for (const auto& t : message.tensors()) {
+      TORCH_CHECK(
+          t.device().is_cpu(),
+          errStr,
+          ", but found tensor on device: ",
+          t.device());
+    }
+    return {};
+  } else {
+    std::vector<c10::DeviceIndex> deviceIndices;
+    deviceIndices.reserve(message.tensors().size());
+    const auto& deviceMap = iter->second;
+    for (const auto& t : message.tensors()) {
+      const auto deviceIter = deviceMap.find(t.device().index());
+      if (deviceIter == deviceMap.end()) {
+        TORCH_CHECK(
+            t.device().is_cpu(),
```

Comment:
Nit: couldn't we first check if the tensor is on CPU and only then, in case it isn't, look its device index up in the map?